File: src/lib/encryption.ts
import crypto from "crypto";

const ALGORITHM = "aes-256-gcm";
const KEY_LENGTH = 32;
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;
const SALT_LENGTH = 16;

let cachedKey: Buffer | null = null;

export function getEncryptionKey(): Buffer {
  if (cachedKey) return cachedKey;

  const hexKey = process.env.ENCRYPTION_KEY;

  if (!hexKey) {
    throw new Error("ENCRYPTION_KEY environment variable is not set");
  }

  if (hexKey.length !== KEY_LENGTH * 2) {
    throw new Error(
      `ENCRYPTION_KEY must be exactly ${KEY_LENGTH * 2} hex characters (${KEY_LENGTH} bytes)`
    );
  }

  cachedKey = Buffer.from(hexKey, "hex");
  return cachedKey;
}

export function encrypt(text: string, key: Buffer): string {
  try {
    const iv = crypto.randomBytes(IV_LENGTH);
    const salt = crypto.randomBytes(SALT_LENGTH);

    const derivedKey = crypto.pbkdf2Sync(key, salt, 100_000, KEY_LENGTH, "sha256");

    const cipher = crypto.createCipheriv(ALGORITHM, derivedKey, iv);
    cipher.setAAD(Buffer.from("mkindayzir-ai-key"));

    const encrypted = Buffer.concat([cipher.update(text, "utf8"), cipher.final()]);
    const authTag = cipher.getAuthTag();

    const parts = [
      salt.toString("base64"),
      iv.toString("base64"),
      encrypted.toString("base64"),
      authTag.toString("base64"),
    ];

    return parts.join(".");
  } catch (error) {
    console.error("Encryption failed:", error);
    throw new Error("Failed to encrypt data");
  }
}

export function decrypt(encrypted: string, key: Buffer): string {
  try {
    const parts = encrypted.split(".");

    if (parts.length !== 4) {
      throw new Error("Invalid encrypted data format");
    }

    const [saltB64, ivB64, encryptedB64, authTagB64] = parts;

    const salt = Buffer.from(saltB64, "base64");
    const iv = Buffer.from(ivB64, "base64");
    const encryptedData = Buffer.from(encryptedB64, "base64");
    const authTag = Buffer.from(authTagB64, "base64");

    const derivedKey = crypto.pbkdf2Sync(key, salt, 100_000, KEY_LENGTH, "sha256");

    const decipher = crypto.createDecipheriv(ALGORITHM, derivedKey, iv);
    decipher.setAuthTag(authTag);
    decipher.setAAD(Buffer.from("mkindayzir-ai-key"));

    const decrypted = Buffer.concat([
      decipher.update(encryptedData),
      decipher.final(),
    ]);

    return decrypted.toString("utf8");
  } catch (error) {
    console.error("Decryption failed:", error);
    throw new Error("Failed to decrypt data");
  }
}

File: backend/app/utils/encryption.py
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes
import base64
import os
from app.config import settings

ALGORITHM = "aes-256-gcm"
KEY_LENGTH = 32
IV_LENGTH = 12
AUTH_TAG_LENGTH = 16
SALT_LENGTH = 16
PBKDF2_ITERATIONS = 100000
AAD = b"mkindayzir-ai-key"

_cached_key: bytes | None = None


def get_encryption_key() -> bytes:
    global _cached_key
    if _cached_key:
        return _cached_key

    hex_key = settings.ENCRYPTION_KEY
    if not hex_key:
        raise ValueError("ENCRYPTION_KEY environment variable is not set")
    if len(hex_key) != KEY_LENGTH * 2:
        raise ValueError(
            f"ENCRYPTION_KEY must be exactly {KEY_LENGTH * 2} hex characters ({KEY_LENGTH} bytes)"
        )
    _cached_key = bytes.fromhex(hex_key)
    return _cached_key


def encrypt(text: str, key: bytes) -> str:
    iv = os.urandom(IV_LENGTH)
    salt = os.urandom(SALT_LENGTH)
    derived_key = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH,
        salt=salt,
        iterations=PBKDF2_ITERATIONS,
    ).derive(key)

    aesgcm = AESGCM(derived_key)
    encrypted = aesgcm.encrypt(iv, text.encode("utf-8"), AAD)
    ciphertext = encrypted[:-AUTH_TAG_LENGTH]
    auth_tag = encrypted[-AUTH_TAG_LENGTH:]

    parts = [
        base64.b64encode(salt).decode("utf-8"),
        base64.b64encode(iv).decode("utf-8"),
        base64.b64encode(ciphertext).decode("utf-8"),
        base64.b64encode(auth_tag).decode("utf-8"),
    ]
    return ".".join(parts)


def decrypt(encrypted: str, key: bytes) -> str:
    parts = encrypted.split(".")
    if len(parts) != 4:
        raise ValueError("Invalid encrypted data format")

    salt = base64.b64decode(parts[0])
    iv = base64.b64decode(parts[1])
    ciphertext = base64.b64decode(parts[2])
    auth_tag = base64.b64decode(parts[3])

    derived_key = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH,
        salt=salt,
        iterations=PBKDF2_ITERATIONS,
    ).derive(key)

    aesgcm = AESGCM(derived_key)
    decrypted = aesgcm.decrypt(iv, ciphertext + auth_tag, AAD)
    return decrypted.decode("utf-8")
